src/Cache/Generator/HeaderCacheKeyGenerator.php

```diff
@@ -0,0 +1,38 @@
+<?php
+
+namespace Http\Client\Common\Plugin\Cache\Generator;
+
+use Psr\Http\Message\RequestInterface;
+
+/**
+ * Generate a cache key by specify what headers you want to vary on.
+ *
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ */
+class HeaderCacheKeyGenerator implements CacheKeyGenerator
+{
+    /**
+     * The header names we should take into account when creating the cache key.
+     *
+     * @var array
+     */
+    private $headerNames;
+
+    /**
+     * @param $headerNames
+     */
+    public function __construct(array $headerNames)
+    {
+        $this->headerNames = $headerNames;
+    }
+
+    public function generate(RequestInterface $request)
+    {
+        $concatenatedHeaders = [];
+        foreach ($this->headerNames as $headerName) {
+            $concatenatedHeaders[] = sprintf(' %s:"%s"', $headerName, $request->getHeaderLine($headerName));
```

Comment:
calling getHeaderLine for a header that is not defined simply returns an empty string. just double-checked in the psr-7 spec. so this is fine